cmd/erasure-server-pool.go

```diff
@@ -47,6 +47,82 @@ import (
 	"github.com/minio/pkg/v3/wildcard"
 )
 
+// MultipartUpload represents multipart upload details to be cached
+type MultipartUpload struct {
+	Bucket   string
+	Object   string
+	UploadID string
+}
+
+// MultipartUploadCacheEntry represents one multipart upload cache entry
+type MultipartUploadCacheEntry struct {
+	Upload    MultipartUpload
+	Timestamp time.Time
+}
+
+// MultipartUploadCache represents a in memory cache to hold multipart upload details
+type MultipartUploadCache struct {
+	// Key would be uploadID as it's unique UUID
+	data  map[string]MultipartUploadCacheEntry
+	mutex sync.RWMutex
+}
+
+// NewMultipartUploadsCache initializes and returns a new Cache
+func NewMultipartUploadsCache() *MultipartUploadCache {
+	cache := &MultipartUploadCache{
+		data: make(map[string]MultipartUploadCacheEntry),
+	}
+	// Start the cleanup process
+	go cache.cleanExpiredEntries()
+	return cache
+}
+
+// Set adds a value to the cache with the current timestamp
+func (c *MultipartUploadCache) Set(key string, value MultipartUpload) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.data[key] = MultipartUploadCacheEntry{
+		Upload:    value,
+		Timestamp: time.Now(),
+	}
+}
+
+// Get retrieves a value from the cache if it has not expired
+func (c *MultipartUploadCache) Get(key string) (MultipartUpload, bool) {
+	expiry := globalAPIConfig.getStaleUploadsExpiry()
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	entry, exists := c.data[key]
+	if !exists || time.Since(entry.Timestamp) > expiry {
+		return MultipartUpload{}, false
+	}
+	return entry.Upload, true
+}
+
+// Has returns true if cache contains the key
+func (c *MultipartUploadCache) Has(key string) bool {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	_, exists := c.data[key]
+	return exists
+}
+
+// cleanExpiredEntries removes expired entries every hour
+func (c *MultipartUploadCache) cleanExpiredEntries() {
+	expiry := globalAPIConfig.getStaleUploadsExpiry()
+	for {
+		time.Sleep(1 * time.Hour)
+		c.mutex.Lock()
+		now := time.Now()
+		for key, entry := range c.data {
+			if now.Sub(entry.Timestamp) > expiry {
+				delete(c.data, key)
+			}
+		}
+		c.mutex.Unlock()
+	}
+}
```

Comment:
I think what we can do is rely on cleanupStaleUploadsOnDisk() to return []string of uploadIDs, this uploadIDs you can work on the delete from the map instead of doing a new goroutine here. 

```
func (er erasureObjects) cleanupStaleUploadsOnDisk(ctx context.Context, disk StorageAPI) {
```

Reply:
```go
func (s *erasureSets) cleanupStaleUploads(ctx context.Context) {
        timer := time.NewTimer(globalAPIConfig.getStaleUploadsCleanupInterval())
        defer timer.Stop()

        for {
                select {
                case <-ctx.Done():
                        return
                case <-timer.C:
                        var wg sync.WaitGroup
                        for _, set := range s.sets {
                                wg.Add(1)
                                go func(set *erasureObjects) {
                                        defer wg.Done()
                                        if set == nil {
                                                return
                                        }
                                        set.cleanupStaleUploads(ctx)
                                }(set)
                        }
                        wg.Wait()

                        // Reset for the next interval
                        timer.Reset(globalAPIConfig.getStaleUploadsCleanupInterval())
                }
        }
}
```